leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/CaConstraintCertificateVerifier.java

```diff
@@ -46,60 +48,51 @@
  *       certificate usage allows both trust anchors and CA certificates,
  *       the certificate might or might not have the basicConstraints
  *       extension present.
- *
+ * </pre>
+ * 
  * For details about Certificate Usage please see:
  * <a href="https://tools.ietf.org/html/rfc6698#section-2.1.1">rfc6698#section-2.1.1</a> - The Certificate Usage Field
  */
-public class CaConstraintCertificateVerifier extends LeshanCertificateVerifierBase {
+public class CaConstraintCertificateVerifier extends BaseCertificateVerifier {
 
-    public CaConstraintCertificateVerifier(Certificate expectedServerCertificate,
-            X509Certificate[] trustedCertificates) {
-        super(expectedServerCertificate, trustedCertificates);
+    private final Certificate caCertificate;
+    private final X509Certificate[] trustedCertificates;
+
+    public CaConstraintCertificateVerifier(Certificate caCertificate, X509Certificate[] trustedCertificates) {
+        Validate.notNull(caCertificate);
+        Validate.notNull(trustedCertificates);
+        Validate.notEmpty(trustedCertificates);
+        this.caCertificate = caCertificate;
+        this.trustedCertificates = trustedCertificates;
     }
 
     @Override
     public void verifyCertificate(CertificateMessage message, DTLSSession session) throws HandshakeException {
         CertPath messageChain = message.getCertificateChain();
 
-        if (messageChain.getCertificates().size() == 0) {
-            AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.BAD_CERTIFICATE,
-                    session.getPeer());
-            throw new HandshakeException("Certificate chain could not be validated : server cert chain is empty",
-                    alert);
-        }
-
-        Certificate receivedServerCertificate = messageChain.getCertificates().get(0);
-        if (!(receivedServerCertificate instanceof X509Certificate)) {
-            AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.UNSUPPORTED_CERTIFICATE,
-                    session.getPeer());
-            throw new HandshakeException("Certificate chain could not be validated - unknown certificate type", alert);
-        }
+        validateCertificateChainNotEmpty(messageChain, session.getPeer());
 
+        // - must do PKIX validation with trustStore
         CertPath certPath = null;
-        if (trustedCertificates != null) {
-            try {
-                // - must do PKIX validation with trustStore
-                certPath = CertPathUtil.validateCertificatePath(false, messageChain, trustedCertificates);
-            } catch (GeneralSecurityException e) {
-                AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.BAD_CERTIFICATE,
-                        session.getPeer());
-                throw new HandshakeException("Certificate chain could not be validated", alert, e);
-            }
+        try {
```

Comment:
I remove the null check and add a validateNotNull in constructor.
If we want to support `null` truststore there is more work to do. (I guess with previous code we could will face an NPE at some line below `List<? extends Certificate> certificates = certPath.getCertificates(); `

Reply:
OK